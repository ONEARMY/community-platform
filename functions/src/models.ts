// tslint:disable no-implicit-dependencies
// Models can be imported from the main package for use here
// NOTE 1 - this requires adjustment main src in package.json
// NOTE 2 - shorthand @OAModels notation defined in tsconfig
import { IDBEndpoint, IDbDoc } from '@OAModels/common.models'

export { IDBEndpoint, IDbDoc }
