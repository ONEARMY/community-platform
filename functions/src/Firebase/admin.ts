/***********************************************************
When running on live the config below pulls all data from 
environment variables.
***********************************************************/
import admin from 'firebase-admin'
import { SERVICE_ACCOUNT_CONFIG } from '../config/config'
const serviceAccount = SERVICE_ACCOUNT_CONFIG
const FIREBASE_CONFIG = JSON.parse(
  process.env.FIREBASE_CONFIG || '{}',
) as IFirebaseConfig
let firebaseApp: admin.app.App
// Use Emulator Config
if (process.env.FUNCTIONS_EMULATOR === 'true') {
  console.log('[Admin Configuration] - emulator')
  // HACK - emulator db doesn't configure correctly with env var for some reason
  // https://github.com/firebase/firebase-tools/issues/2852
  // https://github.com/firebase/firebase-admin-node/issues/1096
  const databaseURL = process.env.FIREBASE_DATABASE_EMULATOR_HOST
  delete process.env.FIREBASE_DATABASE_EMULATOR_HOST
  firebaseApp = admin.initializeApp({
    credential: admin.credential.applicationDefault(),
    databaseURL,
    projectId: process.env.GCLOUD_PROJECT,
  })
}
// Use Service Account Config
else if (serviceAccount) {
  console.log(`[Admin Configuration] - ${FIREBASE_CONFIG.projectId}`)
  const cert = {
    clientEmail: serviceAccount.client_email,
    privateKey: serviceAccount.private_key,
    projectId: serviceAccount.project_id,
  }
  firebaseApp = admin.initializeApp({
    credential: admin.credential.cert(cert),
    databaseURL: FIREBASE_CONFIG.databaseURL,
    projectId: FIREBASE_CONFIG.projectId,
  })
}

// Use Google Default Environment (only works on google server)
else {
  console.log(`[Admin Configuration] - ${FIREBASE_CONFIG.projectId}`)
  admin.initializeApp()
}

export { firebaseApp }
export const firebaseAdmin = admin

// this is automatically populated by firebase
export interface IFirebaseConfig {
  databaseURL: string
  storageBucket: string
  projectId: string
}
