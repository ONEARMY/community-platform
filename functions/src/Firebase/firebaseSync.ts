import rtdb from './realtimeDB'
import * as firestore from './firestoreDB'
import { DBDoc, DB_ENDPOINTS } from '../models'

/*  Functions in this folder are used to sync data between firestore and firebase realtime databases
    The reason for this is to allow large collections to be 'cached' for cheap retrieval
    An example would be the map pins. When a user first visits the map all pins are downloaded
    The previous map had 30,000 pins, so this would cost roughly $0.02c, or $20 per 1000 users
    By pulling the data from firebase realtime instead and only subscribing to updates, this is
    dramatically reduced. For more information about the various database strategies see:
    https://github.com/OneArmyWorld/onearmy/wiki/Backend-Database
*/

// List of endpoints to use during sync operations, taken from common db mapping
const endpoints = Object.values(DB_ENDPOINTS)

export const syncAll = async () => {
  const promises = endpoints.map(async (endpoint) => await sync(endpoint))
  const results = await Promise.all(promises)
  const response = {}
  endpoints.forEach((endpoint, i) => (response[endpoint] = results[i]))
  return response
}

// for given endpoint, query rtdb for all records, determin latest,
// query firestore for newer records, add to rtdb
export const sync = async (endpoint: string) => {
  const existing = await rtdb.get(endpoint)
  const latest =
    existing && Object.keys(existing).length > 1
      ? Object.values<DBDoc>(existing).sort((a, b) => _sortByModified(a, b))[0]
          ._modified
      : ''

  const update = await firestore.db
    .collection(endpoint)
    .where('_modified', '>', latest)
    .get()
  const docs = update.empty ? [] : update.docs.map((d) => d.data())
  const json = {}
  docs.forEach((doc) => (json[doc._id] = doc))
  await rtdb.update(endpoint, json)
  return json
}

function _sortByModified(a: DBDoc, b: DBDoc) {
  return a._modified > b._modified ? -1 : 1
}
