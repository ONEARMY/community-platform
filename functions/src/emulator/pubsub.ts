import { PubSub } from '@google-cloud/pubsub'

interface ITriggerOptions {
  trigger: 'topic' | 'schedule'
  name: string
  payload: any
}
const defaultOptions: ITriggerOptions = {
  trigger: 'topic',
  name: 'test-topic',
  payload: {},
}

/**
 * Trigger a pubsub message
 * @param options.trigger - specify as 'schedule' if trying to call a scheduled function.
 * Default 'topic'
 * @param options.name - name of the topic to be scheduled. This will either be the name of the
 * scheduled function, or the name of the pubsub topic to trigger. Default "test-topic"
 * @param options.payload - optional payload passed to function. Default {}
 * @returns confirmation messageID of triggered pubsub
 */
export const triggerPubsub = async (options: ITriggerOptions) => {
  let { name, payload, trigger } = { ...defaultOptions, ...options }
  const pubsub = new PubSub()
  if (trigger === 'schedule') {
    // scheduled functions can be triggered by adding prefix to topic name
    // https://github.com/firebase/firebase-tools/pull/2011/files
    name = `firebase-schedule-${name}`
  }

  const [topics] = await pubsub.getTopics()
  if (!topics.find((t) => t.name.includes(name))) {
    await pubsub.createTopic(name)
  }

  const messageID = await pubsub.topic(name).publishMessage({ json: payload })
  return messageID
}
