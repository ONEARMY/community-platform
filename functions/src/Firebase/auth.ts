import { getAuth } from 'firebase-admin/auth'
import { firebaseApp } from './admin'

export const firebaseAuth = getAuth(firebaseApp)
