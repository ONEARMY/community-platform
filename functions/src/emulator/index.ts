import * as functions from 'firebase-functions'
import cors from 'cors'
import express from 'express'
import { seedDataClean } from './seed/data-clean'
import { seedUsersCreate } from './seed/users-create'
import { triggerPubsub } from './pubsub'

console.log('emulator api ready')

const app = express()
app.use(cors({ origin: true }))

/**
 * Expose an api endpoint available just for use with emulators
 * NOTE - if changing endpoints check emulators workspace in case in use (e.g. called via Dockerfile)
 */
app.use(function (req, res, next) {
  if (process.env.FUNCTIONS_EMULATOR === 'true') next()
  else
    res
      .status(403)
      .send(
        `Emulator api methods can only be accessed in emulator environment [${process.env.FUNCTIONS_EMULATOR}]`,
      )
})

app.get('/', (req, res) => res.status(200).send('Emulator Api Working'))

app.get('/seed-users-create', (req, res) =>
  seedUsersCreate().then((users) =>
    res
      .status(200)
      .send(`Created users: ${users.map((u) => u.uid).join(', ')}`),
  ),
)

app.get('/seed-clean', (req, res) =>
  seedDataClean().then((result) => {
    res.status(200).send(result)
  }),
)

/**
 * As there is currently no UI to publish pubsub messages for testing, create
 * a http wrapper that can be used to call via rest
 * https://github.com/firebase/firebase-tools/issues/2034
 * https://blog.minimacode.com/publish-message-to-pubsub-emulator/
 *
 * This allows pubsub topics to be tested when running in emulators as an api endpoint
 * It includes additional prefix (/topic or /schedule) depending on trigger
 * @example
 * `http://localhost:4002/community-platform-emulated/us-central1/emulator/pubsub/schedule/dailyTasks`
 * `http://localhost:4002/community-platform-emulated/us-central1/emulator/pubsub/topic/test-topic`
 */
app.all('/pubsub/:trigger/:name', async (req, res) => {
  if (!process.env.PUBSUB_EMULATOR_HOST) {
    functions.logger.error(
      'This function should only run locally in an emulator.',
    )
    res.status(400).end()
  }
  const trigger = req.params.trigger as 'topic' | 'schedule'
  const name = req.params.name
  const payload = req.body
  const messageId = await triggerPubsub({ trigger, name, payload })
  const result = 'Pubsub Published'
  res.status(201).send({ result, trigger, name, messageId, payload })
})

// Increase default timeout to max allowed for longer-running operations
export = functions.runWith({ timeoutSeconds: 540 }).https.onRequest(app as any)
