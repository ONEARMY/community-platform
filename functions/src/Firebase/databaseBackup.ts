import Axios from 'axios'
import { getAccessToken } from '../Utils/auth.utils'
import { SERVICE_ACCOUNT_CONFIG } from '../config/config'
import * as google from 'google-auth-library'
import * as dateformat from 'dateformat'

/*  Cloud function to automatically backup the firebase database adapted from: 
    https://thatweirddeveloper.com/how-to-back-up-cloud-firestore-periodically

    More examples at:
    https://firebase.google.com/docs/firestore/solutions/schedule-export

    Note this requires use of a service account to access drive storage,
    which is accessed from environment variables
*/

// rest reference: https://cloud.google.com/firestore/docs/reference/rest/v1beta2/projects.databases/exportDocuments
export const BackupDatabase = async () => {
  console.log('executing database backup')
  // TODO - no longer using util auth, should revise code to see if still relevant or not
  const auth = await google.auth.getClient({
    scopes: ['https://www.googleapis.com/auth/datastore'],
  })
  const accessTokenResponse = await auth.getAccessToken()
  const accessToken = accessTokenResponse.token
  const headers = {
    'Content-Type': 'application/json',
    Authorization: 'Bearer ' + accessToken,
  }

  console.log('access token received', accessToken)
  const PROJECT_ID = SERVICE_ACCOUNT_CONFIG.project_id

  const url = `https://firestore.googleapis.com/v1beta1/projects/${PROJECT_ID}/databases/(default):exportDocuments`
  // use axios to send post request as promise
  console.log('posting', url)
  const timestamp = dateformat(Date.now(), 'yyyy-mm-dd-HH-MM-ss')
  const body = {
    outputUriPrefix: `gs://${PROJECT_ID}.appspot.com/backups/${timestamp}`,
  }
  try {
    const response = await Axios.post(url, body, { headers: headers })
    return { status: 200, message: response.data }
  } catch (error) {
    return { status: 500, message: error.response.data }
  }
}

// const createSub = async () => {
//   pubsub
//     .topic('firebase-backup')
//     .onPublish(async (message, context) => BackupDatabase)
// }
