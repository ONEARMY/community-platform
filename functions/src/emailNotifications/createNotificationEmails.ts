import { EmailNotificationFrequency } from 'oa-shared'
import { INotification } from '../../../src/models'
import { firebaseAuth } from '../Firebase/auth'
import { db } from '../Firebase/firestoreDB'
import { DB_ENDPOINTS, IUserDB, IPendingEmails } from '../models'
import { getNotificationEmailTemplate } from './getNotificationEmailTemplate'

const getUserEmail = async (uid: string): Promise<string | null> => {
  try {
    const { email } = await firebaseAuth.getUser(uid)
    return email
  } catch (error) {
    return null
  }
}

const updateEmailedNotifications = async (
  user: FirebaseFirestore.DocumentSnapshot<IUserDB>,
  emailedNotifications: INotification[],
  emailField: string,
) => {
  const updatedNotifications = user.data().notifications.map((notification) => {
    if (
      emailedNotifications.some(
        (emailedNotification) => (emailedNotification._id = notification._id),
      )
    ) {
      return {
        ...notification,
        email: emailField,
      }
    }
    return notification
  })

  await user.ref
    .update({ notifications: updatedNotifications })
    .catch((error) => {
      throw new Error(
        `Error updating user notifications: ${JSON.stringify(error)}`,
      )
    })
}

// Create an email from a set of pending notifications for a user.
const handlePendingEmailEntry = async (
  userName: string,
  { notifications: pendingNotifications, emailFrequency }: IPendingEmails,
  targetFrequency?: EmailNotificationFrequency,
) => {
  if (
    !pendingNotifications.length ||
    (targetFrequency !== undefined && emailFrequency !== targetFrequency) ||
    emailFrequency === EmailNotificationFrequency.NEVER
  ) {
    return
  }

  const toUserDoc = (await db
    .collection(DB_ENDPOINTS.users)
    .doc(userName)
    .get()) as FirebaseFirestore.DocumentSnapshot<IUserDB>

  const toUser = toUserDoc.exists ? toUserDoc.data() : undefined
  const toUserEmail = toUser ? await getUserEmail(toUser._authID) : undefined

  if (!toUser || !toUserEmail) {
    await updateEmailedNotifications(toUserDoc, pendingNotifications, 'failed')
    throw new Error(`Cannot get user ${userName}`)
  }

  try {
    // Adding emails to this collection triggers an email notification to be sent to the user
    const sentEmailRef = await db.collection(DB_ENDPOINTS.emails).add({
      to: toUserEmail,
      message: getNotificationEmailTemplate(toUser, pendingNotifications),
    })

    await updateEmailedNotifications(
      toUserDoc,
      pendingNotifications,
      sentEmailRef.id,
    )
  } catch (error) {
    await updateEmailedNotifications(toUserDoc, pendingNotifications, 'failed')
    throw new Error(
      `Error sending an email to ${userName}: ${JSON.stringify(error)}`,
    )
  }
}

export async function createNotificationEmails(
  targetFrequency?: EmailNotificationFrequency,
) {
  const pendingEmails = await db
    .collection(DB_ENDPOINTS.user_notifications)
    .doc('emails_pending')
    .get()

  const pendingEmailEntries =
    Object.entries<IPendingEmails>(pendingEmails.data()) ?? []

  for (const entry of pendingEmailEntries) {
    await handlePendingEmailEntry(...entry, targetFrequency)
  }
}
