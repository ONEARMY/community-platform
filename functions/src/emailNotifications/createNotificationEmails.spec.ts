import { EmailNotificationFrequency } from 'oa-shared'

import { DB_ENDPOINTS } from '../models'
import { FirebaseEmulatedTest } from '../test/Firebase/emulator'
import { createNotificationEmails } from './createNotificationEmails'

import type { INotification, IUserDB } from '../models'

jest.mock('../Firebase/auth', () => ({
  firebaseAuth: {
    getUser: () => ({
      email: 'test@test.com',
    }),
  },
}))

jest.mock('../config/config', () => ({
  CONFIG: {
    deployment: {
      site_url: 'https://community.preciousplastic.com',
    },
  },
}))

jest.mock('uuid', () => ({
  v4: () => 'test',
}))

const notificationFactory = (
  _userId: string,
  _id: string,
  notification: Partial<INotification>,
): INotification => ({
  _id,
  _created: '',
  triggeredBy: {
    displayName: '',
    userId: '',
  },
  relevantUrl: '',
  type: 'new_comment_discussion',
  read: false,
  notified: false,
  ...notification,
})

const userFactory = (_id: string, user: Partial<IUserDB> = {}): IUserDB => ({
  _id,
  _authID: _id,
  ...(user as IUserDB),
})

describe('create email test', () => {
  const db = FirebaseEmulatedTest.admin.firestore()

  beforeEach(async () => {
    await FirebaseEmulatedTest.clearFirestoreDB()
    await FirebaseEmulatedTest.seedFirestoreDB('emails')

    const user1Notifications = [
      notificationFactory('user_1', 'notification_1', {
        triggeredBy: {
          displayName: 'User 2',
          userId: 'user_2',
        },
        relevantUrl: '/test',
        type: 'howto_useful',
      }),
      notificationFactory('user_1', 'notification_2', {
        triggeredBy: {
          displayName: 'User 2',
          userId: 'user_2',
        },
        relevantUrl: '/test',
        title: 'Test research',
        type: 'new_comment_discussion',
      }),
      notificationFactory('user_1', 'notification_3', {
        triggeredBy: {
          displayName: 'User 3',
          userId: 'user_3',
        },
        relevantUrl: '/test',
        type: 'howto_mention',
      }),
      notificationFactory('user_1', 'notification_4', {
        triggeredBy: {
          displayName: 'User 1',
          userId: 'user_1',
        },
        relevantUrl: '/test',
        type: 'map_pin_approved',
      }),
    ]

    const user2Notifications = [
      notificationFactory('user_2', 'notification_1', {
        triggeredBy: {
          displayName: 'User 1',
          userId: 'user_1',
        },
        relevantUrl: '/test',
        title: 'Test how-to',
        type: 'new_comment_discussion',
      }),
      notificationFactory('user_2', 'notification_2', {
        triggeredBy: {
          displayName: 'User 3',
          userId: 'user_3',
        },
        relevantUrl: '/test',
        title: 'Test research',
        type: 'new_comment_discussion',
      }),
      notificationFactory('user_2', 'notification_3', {
        triggeredBy: {
          displayName: 'User 3',
          userId: 'user_3',
        },
        relevantUrl: '/test',
        type: 'research_update',
      }),
      notificationFactory('user_2', 'notification_4', {
        triggeredBy: {
          displayName: 'User 2',
          userId: 'user_2',
        },
        relevantUrl: '/test',
        type: 'research_needs_updates',
      }),
    ]

    const user3Notifications = [
      notificationFactory('user_3', 'notification_1', {
        triggeredBy: {
          displayName: 'User 1',
          userId: 'user_1',
        },
        relevantUrl: '/test',
        title: 'Test how-to',
        type: 'new_comment_discussion',
      }),
    ]

    await FirebaseEmulatedTest.seedFirestoreDB('users', [
      userFactory('user_1', {
        notifications: user1Notifications,
        displayName: 'User 1',
        userName: 'user_1',
      }),
      userFactory('user_2', {
        notifications: user2Notifications,
        displayName: 'User 2',
        userName: 'user_2',
      }),
      userFactory('user_3', {
        notifications: user3Notifications,
        displayName: 'User 3',
        userName: 'user_3',
      }),
      userFactory('user_4', {
        displayName: 'User 4',
        userName: 'user_4',
      }),
    ])

    await FirebaseEmulatedTest.seedFirestoreDB('user_notifications')

    await db
      .collection(DB_ENDPOINTS['user_notifications'])
      .doc('emails_pending')
      .set({
        ['user_1']: {
          _authId: '',
          _userId: 'user_1',
          notifications: user1Notifications,
          emailFrequency: EmailNotificationFrequency.WEEKLY,
        },
        ['user_2']: {
          _authId: '',
          _userId: 'user_2',
          notifications: user2Notifications,
          emailFrequency: EmailNotificationFrequency.MONTHLY,
        },
        ['user_3']: {
          _authId: '',
          _userId: 'user_3',
          notifications: user3Notifications,
          emailFrequency: EmailNotificationFrequency.NEVER,
        },
        ['user_4']: {
          _authId: '',
          _userId: 'user_4',
          notifications: [],
          emailFrequency: EmailNotificationFrequency.DAILY,
        },
      })
  })

  afterEach(async () => {
    await FirebaseEmulatedTest.clearFirestoreDB()
  })

  it('Creates email from pending notifications weekly', async () => {
    await createNotificationEmails(EmailNotificationFrequency.WEEKLY)

    // Only one weekly email should have been created
    const countSnapshot = await db.collection(DB_ENDPOINTS.emails).count().get()
    expect(countSnapshot.data().count).toEqual(1)

    const querySnapshot = await db.collection(DB_ENDPOINTS.emails).get()

    querySnapshot.forEach((doc) => {
      const {
        message: { html, subject },
        to,
      } = doc.data()
      expect(html).toMatchSnapshot()
      expect(subject).toBe(`You've missed notifications from Precious Plastic`)
      expect(to).toBe('test@test.com')
    })

    // Notifications should have been updated with email id
    const notificationsQuerySnapshot = await db
      .collection(DB_ENDPOINTS.users)
      .doc('user_1')
      .get()
    const { notifications } = notificationsQuerySnapshot.data()
    notifications.forEach(({ email }) =>
      expect(email).toBe(querySnapshot.docs[0].id),
    )
  })

  it('Creates email from pending notifications monthly', async () => {
    await createNotificationEmails(EmailNotificationFrequency.MONTHLY)

    // Only one monthly email should have been created
    const countSnapshot = await db.collection(DB_ENDPOINTS.emails).count().get()
    expect(countSnapshot.data().count).toEqual(1)

    const querySnapshot = await db.collection(DB_ENDPOINTS.emails).get()
    querySnapshot.forEach((doc) => {
      const {
        message: { html, subject },
        to,
      } = doc.data()
      expect(html).toMatchSnapshot()
      expect(subject).toBe(`You've missed notifications from Precious Plastic`)
      expect(to).toBe('test@test.com')
    })

    // Notifications should have been updated with email id
    const notificationsQuerySnapshot = await db
      .collection(DB_ENDPOINTS.users)
      .doc('user_2')
      .get()
    const { notifications } = notificationsQuerySnapshot.data()
    notifications.forEach(({ email }) =>
      expect(email).toBe(querySnapshot.docs[0].id),
    )
  })

  it('Creates emails from pending notifications all', async () => {
    await createNotificationEmails()

    // Two emails should have been created
    const countSnapshot = await db.collection(DB_ENDPOINTS.emails).count().get()
    expect(countSnapshot.data().count).toEqual(2)
  })
})
