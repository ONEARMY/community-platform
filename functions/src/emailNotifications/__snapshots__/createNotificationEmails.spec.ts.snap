// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create email test Creates email from pending notifications monthly 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html charset=UTF-8" />
    <style>
      #body {
        font-family: Arial, Helvetica, sans-serif;
        font-size: 14px;
        color: #000000;
      }
      #body a {
        color: inherit;
      }
      .border {
        border: 3px solid black;
        border-radius: 10px;
        padding: 4% 0;
        margin: 4% 0;
        width: 550px;
      }
      table {
        width: 100%;
        border: 0;
      }
      .settings-table-container,
      .project-image-table-container,
      .greeting-container {
        margin-bottom: 8%;
      }
      .greeting-container,
      .notifications-container {
        margin-left: 8%;
      }
      .notifications-header {
        font-size: 24px;
      }
      .notifications-header-icon {
        margin-right: 8px;
      }
      @media only screen and (max-width: 550px) {
        .table-container {
          width: 100%;
        }
      }
    </style>
  </head>
  <body id="body">
    <table class="email-table-container">
      <tr>
        <td align="center">
          <div class="border">
            <table class="project-image-table-container">
              <tr>
                <td align="center">
                  <img
                    width="144"
                    alt=""
                    src="http://cdn.mcauto-images-production.sendgrid.net/cb5a685b085b7e7c/f48e2455-6dbb-4102-bde9-2ee07d3008c4/859x860.png"
                  />
                </td>
              </tr>
            </table>
            <div align="left" class="greeting-container">
              <p>Hey User 2</p>
              <p>You've missed notifications. No worries.</p>
              <p>I'm here to give you an overview :)</p>
            </div>
            <div align="left" class="notifications-container">
              <div class="notifications-header">
                <img
                  class="notifications-header-icon"
                  width="30"
                  height="24"
                  alt=""
                  src="http://cdn.mcauto-images-production.sendgrid.net/cb5a685b085b7e7c/1e6f3be3-e6fd-46c5-b30f-bcde950b0dbc/165x131.png"
                />
                Missed Notifications
              </div>

              <p>
                New comment by
                <a href="https://community.preciousplastic.com/u/user_1"
                  >User 1</a
                >
                on
                <a href="https://community.preciousplastic.com/test"
                  >Test how-to</a
                >
              </p>
              <p>
                New comment by
                <a href="https://community.preciousplastic.com/u/user_3"
                  >User 3</a
                >
                on
                <a href="https://community.preciousplastic.com/test"
                  >Test research</a
                >
              </p>
              <p>
                <a href="https://community.preciousplastic.com/u/user_3"
                  >User 3</a
                >
                posted an update to this
                <a href="https://community.preciousplastic.com/test"
                  >research</a
                >
                you follow
              </p>

              <p>
                Your
                <a href="https://community.preciousplastic.com/test"
                  >research</a
                >
                needs updates
              </p>
            </div>
          </div>
          <table class="settings-table-container">
            <tr>
              <td align="center">
                <p>
                  Manage your notifications
                  <a href="https://community.preciousplastic.com/settings">
                    here</a
                  >
                </p>
                <p>
                  <a
                    href="https://community.preciousplastic.com/unsubscribe/test"
                    >Unsubscribe</a
                  >
                </p>
              </td>
            </tr>
          </table>
        </td>
      </tr>
    </table>
  </body>
</html>
"
`;

exports[`create email test Creates email from pending notifications weekly 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html charset=UTF-8" />
    <style>
      #body {
        font-family: Arial, Helvetica, sans-serif;
        font-size: 14px;
        color: #000000;
      }
      #body a {
        color: inherit;
      }
      .border {
        border: 3px solid black;
        border-radius: 10px;
        padding: 4% 0;
        margin: 4% 0;
        width: 550px;
      }
      table {
        width: 100%;
        border: 0;
      }
      .settings-table-container,
      .project-image-table-container,
      .greeting-container {
        margin-bottom: 8%;
      }
      .greeting-container,
      .notifications-container {
        margin-left: 8%;
      }
      .notifications-header {
        font-size: 24px;
      }
      .notifications-header-icon {
        margin-right: 8px;
      }
      @media only screen and (max-width: 550px) {
        .table-container {
          width: 100%;
        }
      }
    </style>
  </head>
  <body id="body">
    <table class="email-table-container">
      <tr>
        <td align="center">
          <div class="border">
            <table class="project-image-table-container">
              <tr>
                <td align="center">
                  <img
                    width="144"
                    alt=""
                    src="http://cdn.mcauto-images-production.sendgrid.net/cb5a685b085b7e7c/f48e2455-6dbb-4102-bde9-2ee07d3008c4/859x860.png"
                  />
                </td>
              </tr>
            </table>
            <div align="left" class="greeting-container">
              <p>Hey User 1</p>
              <p>You've missed notifications. No worries.</p>
              <p>I'm here to give you an overview :)</p>
            </div>
            <div align="left" class="notifications-container">
              <div class="notifications-header">
                <img
                  class="notifications-header-icon"
                  width="30"
                  height="24"
                  alt=""
                  src="http://cdn.mcauto-images-production.sendgrid.net/cb5a685b085b7e7c/1e6f3be3-e6fd-46c5-b30f-bcde950b0dbc/165x131.png"
                />
                Missed Notifications
              </div>

              <p>
                <a href="https://community.preciousplastic.com/u/user_2"
                  >User 2</a
                >
                found your
                <a href="https://community.preciousplastic.com/test">how-to</a>
                useful
              </p>

              <p>
                New comment by
                <a href="https://community.preciousplastic.com/u/user_2"
                  >User 2</a
                >
                on
                <a href="https://community.preciousplastic.com/test"
                  >Test research</a
                >
              </p>
              <p>
                <a href="https://community.preciousplastic.com/u/user_3"
                  >User 3</a
                >
                mentioned you in this
                <a href="https://community.preciousplastic.com/test">how-to</a>
              </p>
              <p>
                Your
                <a href="https://community.preciousplastic.com/test">map pin</a>
                has been approved
              </p>
            </div>
          </div>
          <table class="settings-table-container">
            <tr>
              <td align="center">
                <p>
                  Manage your notifications
                  <a href="https://community.preciousplastic.com/settings">
                    here</a
                  >
                </p>
                <p>
                  <a
                    href="https://community.preciousplastic.com/unsubscribe/test"
                    >Unsubscribe</a
                  >
                </p>
              </td>
            </tr>
          </table>
        </td>
      </tr>
    </table>
  </body>
</html>
"
`;
